src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -2333,192 +1455,108 @@ class HandshakeForDeflation : public HandshakeClosure {
-                             Atomic::load(&om_list_globals._free_count),
-                             Atomic::load(&om_list_globals._wait_count));
-
-  GVars.stw_random = os::random();
-
   if (self->is_Java_thread()) {
     // The async deflation request has been processed.
     _last_async_deflation_time_ns = os::javaTimeNanos();
     set_is_async_deflation_requested(false);
   }
 
-  if (Atomic::load(&om_list_globals._wait_count) > 0) {
-    // There are deflated ObjectMonitors waiting for a handshake
-    // (or a safepoint) for safety.
+  LogStreamHandle(Debug, monitorinflation) lsh_debug;
+  LogStreamHandle(Info, monitorinflation) lsh_info;
+  LogStream* ls = NULL;
+  if (log_is_enabled(Debug, monitorinflation)) {
+    ls = &lsh_debug;
+  } else if (log_is_enabled(Info, monitorinflation)) {
+    ls = &lsh_info;
+  }
 
-    ObjectMonitor* list = Atomic::load(&om_list_globals._wait_list);
-    assert(list != NULL, "om_list_globals._wait_list must not be NULL");
-    int count = Atomic::load(&om_list_globals._wait_count);
-    Atomic::store(&om_list_globals._wait_count, 0);
-    OrderAccess::storestore();  // Make sure counter update is seen first.
-    Atomic::store(&om_list_globals._wait_list, (ObjectMonitor*)NULL);
+  elapsedTimer timer;
+  if (ls != NULL) {
+    ls->print_cr("begin deflating: in_use_list stats: ceiling=" SIZE_FORMAT ", count=" SIZE_FORMAT ", max=" SIZE_FORMAT,
+                 in_use_list_ceiling(), _in_use_list.count(), _in_use_list.max());
+    timer.start();
+  }
 
-    // Find the tail for prepend_list_to_common(). No need to mark
-    // ObjectMonitors for this list walk since only the deflater
-    // thread manages the wait list.
-#ifdef ASSERT
-    int l_count = 0;
-#endif
-    ObjectMonitor* tail = NULL;
-    for (ObjectMonitor* n = list; n != NULL; n = unmarked_next(n)) {
-      tail = n;
-#ifdef ASSERT
-      l_count++;
-#endif
-    }
-    assert(count == l_count, "count=%d != l_count=%d", count, l_count);
+  // Deflate some idle ObjectMonitors.
+  size_t deflated_count = deflate_monitor_list(self, ls, &timer);
+  if (deflated_count > 0 || is_final_audit()) {
+    // There are ObjectMonitors that have been deflated or this is the
+    // final audit and all the remaining ObjectMonitors have been
+    // deflated, BUT the MonitorDeflationThread blocked for the final
+    // safepoint during unlinking.
 
+    // Unlink deflated ObjectMonitors from the in-use list.
+    ResourceMark rm;
+    GrowableArray<ObjectMonitor*> delete_list((int)deflated_count);
+    size_t unlinked_count = _in_use_list.unlink_deflated(self, ls, &timer,
+                                                         &delete_list);
     if (self->is_Java_thread()) {
+      if (ls != NULL) {
+        timer.stop();
+        ls->print_cr("before handshaking: unlinked_count=" SIZE_FORMAT ", in_use_list stats: ceiling=" SIZE_FORMAT ", count=" SIZE_FORMAT ", max=" SIZE_FORMAT,
```

Comment:
Style nit: line too long

Reply:
Yeah... I keep going back and forth for long logging lines...
I reformatted that one and another one a few lines farther down.